### ISSUE TITLE
UGen math structure (a+b) + (a+b) is inappropriately deleted by the optimizer

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#5047]
## Environment

* SuperCollider version: 3.11
* Operating system: Irrelevant
* Other details (Qt version, audio driver, etc.): Irrelevant

## Steps to reproduce

```supercollider
SynthDef(\test1, {
    var sig = SinOsc.kr(0.1) + Rand(-0.02, 0.02);
    Out.kr(0, sig);
}).dumpUGens;

test1
[ 0_SinOsc, control, [ 0.1, 0.0 ] ]
[ 1_Rand, scalar, [ -0.02, 0.02 ] ]
[ 2_+, control, [ 0_SinOsc, 1_Rand ] ]
[ 3_Out, control, [ 0, 2_+ ] ]

SynthDef(\test2, {
    var sig = SinOsc.kr(0.1) + Rand(-0.02, 0.02);
    Out.kr(0, sig + sig);
}).dumpUGens;

test2
[ 0_SinOsc, control, [ 0.1, 0.0 ] ]
[ 1_Rand, scalar, [ -0.02, 0.02 ] ]
```

(Note: The UGen rates don't matter here -- Rand is `ir` but I tested `kr` as well, same problem.)

## Expected vs. actual behavior

The structure of `\test2` should be:

1. SinOsc
2. Rand
3. '+' (ugen1, ugen2)
4. '+' (ugen3, ugen3)
5. Out (ugen4)

But the binary math operator optimization is getting confused and deleting both operations and their descendent(s).

I haven't done any analysis beyond this.

Once the reason is found, the solution should be apparent -- but this is a complex part of the code, so the analysis is likely to take a while.

**Comment by jamshark70** [comment]
> ... the analysis is likely to take a while.

Or, not that long.

The problem occurs when a `+` UGen plugs into more than one input *of the same `+` unit*.

`optimizeToSum3` is supposed to take `a + b + c` and turn it into `Sum3(a, b, c)` -- removing `a + b` from the graph. This is valid only when `a + b` is plugged into one and only one place (the first input of `something + c`. If `a + b` is plugged into any other input, anywhere, then we need the result of `a + b` as an independent calculation -- we can't fold it entirely into Sum3 -- so the optimization is supposed to be canceled at that point.

We check for this cancellation case by `a.descendants.size == 1`.

But... `a.descendants` is a collection of unique UGen objects.

This is a corner case where, in `something + c`, `a + b` plugs into *both* `something` and `c`. `a + b` properly has two descendents, but they are the same, identical UGen. Under the current rule, this single descendent UGen should be represented in the array only once. So it looks like a valid optimization, but it isn't.

Possible solutions:

- Hack: If `c = a + b`, then the UGen being optimized is `c + c`. So we can check that the two inputs are identical and skip the optimization if true. (Actually, we could do one better: `Sum4(a.inputs[0], a.inputs[0], a.inputs[1], a.inputs[1])` because `(a + b) + (a + b) == (a + a + b + b)`.) Con: Leaving poor architecture in place.

- UGen could have an array of `descendantInputs` in addition to `descendants` and use the array of inputs for checking valid optimizations. Con: Maintenance burden.

- Or, change `descendants` to be unique per input. Con: We don't know how many places in the code rely on objects not to be duplicated in this array -- gut feeling that this would require a *ton* of regression testing.

The hack solution would look like this:

```
	optimizeToSum3 {
		var a, b, replacement;
		#a, b = inputs;
		if(a.rate == \demand or: { b.rate == \demand }) { ^nil };

		if (a.isKindOf(BinaryOpUGen) and: { a.operator == '+'
			and: { a.descendants.size == 1 }}) {
			buildSynthDef.removeUGen(a);
			if(a === b) {
				replacement = Sum4(a.inputs[0], a.inputs[0], a.inputs[1], a.inputs[1]);
			} {
				replacement = Sum3(a.inputs[0], a.inputs[1], b).descendants_(descendants);
			};
			this.optimizeUpdateDescendants(replacement, a);
			^replacement;
		};

		if (b.isKindOf(BinaryOpUGen) and: { b.operator == '+'
			and: { b.descendants.size == 1 }}) {
			buildSynthDef.removeUGen(b);
			if(a === b) {
				replacement = Sum4(b.inputs[0], b.inputs[0], b.inputs[1], b.inputs[1]);
			} {
				replacement = Sum3(b.inputs[0], b.inputs[1], a).descendants_(descendants);
			};
			this.optimizeUpdateDescendants(replacement, b);
			^replacement;
		};
		^nil
	}
```

... with the result:

```
SynthDef(\test2, {
	var sig = SinOsc.kr(0.1) + Rand(-0.02, 0.02);
	Out.kr(0, sig + sig);
}).dumpUGens;

->

test2
[ 0_SinOsc, control, [ 0.1, 0.0 ] ]
[ 1_Rand, scalar, [ -0.02, 0.02 ] ]
[ 2_Sum4, control, [ 0_SinOsc, 0_SinOsc, 1_Rand, 1_Rand ] ]
[ 3_Out, control, [ 0, 2_Sum4 ] ]
```

**Comment by jamshark70** [comment]
I added a PR for the "hack" fix, because I think the Sum4 optimization is a valuable special case to keep.

**Comment by brianlheim** [comment]
fixed in #5047, thanks @jamshark70!
